### PR TITLE
Stereo3d presets

### PR DESCRIPTION
**Pull request by void-ghost**
This adds customizable stereo-3d presets.
Added 3 depth/convergence presets. These are adjustable via (existing) hotkeys - changes to depth and convergence are applied to current preset.
Added 3 new hotkeys for activating presets and a hotkey for toggle between first and second preset.
Added OSD message to convergence/depth changes.
Presets are saved into per-game configs.
Changed depth/convergence hotkeys from "onHold" to "onPress" for more precise control.

Also added getter for video config filename (and removed repeated definitions for that filename). That was needed for first implementation, that saved presets into video config. This change is not needed by current implementation (but is good nonetheless, in my opinion).

### CODE REVIEW

**Comment by RisingFog** [comment]
@dolphin-emu-bot rebuild

File: Source/Core/Core/HotkeyManager.cpp [hunk 1]
```diff
@@ -116,6 +116,12 @@ const std::string hotkey_labels[] =
 	_trans("Undo Save State"),
 	_trans("Save State"),
 	_trans("Load State"),
+
+	_trans("Toggle 3d Preset"),
+	_trans("Use 3d Preset 1"),
+	_trans("Use 3d Preset 2"),
+	_trans("Use 3d Preset 3"),
+
```
Comment (mbc07) [inline Source/Core/Core/HotkeyManager.cpp]: 3d => 3D

File: Source/Core/VideoBackends/D3D/main.cpp [hunk 2]
```diff
@@ -63,7 +63,10 @@ std::string VideoBackend::GetDisplayName() const
 {
 	return "Direct3D";
 }
-
+std::string VideoBackend::GetConfigName() const
+{
+	return "gfx_dx11";
+}
```
Comment (phire) [inline Source/Core/VideoBackends/D3D/main.cpp]: Newline before and after function.

File: Source/Core/VideoBackends/D3D/main.cpp [hunk 3]
```diff
@@ -161,7 +164,7 @@ bool VideoBackend::Initialize(void *window_handle)
 
 	frameCount = 0;
 
-	g_Config.Load(File::GetUserPath(D_CONFIG_IDX) + "gfx_dx11.ini");
+	g_Config.Load(File::GetUserPath(D_CONFIG_IDX) + GetConfigName()+".ini");
```
Comment (phire) [inline Source/Core/VideoBackends/D3D/main.cpp]: spaces around +

File: Source/Core/VideoBackends/OGL/main.cpp [hunk 4]
```diff
@@ -96,6 +96,10 @@ std::string VideoBackend::GetDisplayName() const
 	else
 		return "OpenGL";
 }
+std::string VideoBackend::GetConfigName() const
```
Comment (phire) [inline Source/Core/VideoBackends/OGL/main.cpp]: Newline before function

File: Source/Core/VideoCommon/VideoConfig.h [hunk 5]
```diff
@@ -51,6 +51,14 @@ enum StereoMode
 	STEREO_3DVISION
 };
 
+constexpr int STEREOSCOPY_PRESETS_NUM = 3;
```
Comment (degasus) [inline Source/Core/VideoCommon/VideoConfig.h]: Nice idea. But why exactly 3? We can also just store a vector :)

But so the hotkeys would likely be HK_USE_STEREOSCOPY_PRESET_INC and HK_USE_STEREOSCOPY_PRESET_DEC.

Feel free to ignore this comment.

File: Source/Core/VideoCommon/VideoConfig.cpp [hunk 6]
```diff
@@ -96,6 +96,20 @@ void VideoConfig::Load(const std::string& ini_file)
 	enhancements->Get("StereoConvergence", &iStereoConvergence, 20);
 	enhancements->Get("StereoSwapEyes", &bStereoSwapEyes, false);
 
+	//currently these settings are not saved in global config, so we could've initialized them directly
+	for (int i = 0; i < STEREOSCOPY_PRESETS_NUM; ++i)
+	{
+		int val=iStereoConvergence;
```
Comment (degasus) [inline Source/Core/VideoCommon/VideoConfig.cpp]: spaces around " = "

File: Source/Core/VideoCommon/VideoConfig.cpp [hunk 7]
```diff
@@ -96,6 +96,20 @@ void VideoConfig::Load(const std::string& ini_file)
...
+		enhancements->Get(StringFromFormat("StereoConvergence_%d", i), &val, val);
+		oStereoPresets[i].depth = val;
+		val = iStereoDepth;
+		enhancements->Get(StringFromFormat("StereoDepth_%d", i), &val, val);
+		oStereoPresets[i].convergence = val;
```
Comment (degasus) [inline Source/Core/VideoCommon/VideoConfig.cpp]: No need for a temporary variable:
```
enhancements->Get(StringFromFormat("StereoConvergence_%d", i), &oStereoPresets[i].depth, iStereoConvergence);
enhancements->Get(StringFromFormat("StereoDepth_%d", i), &oStereoPresets[i].convergence, iStereoDepth);
```

File: Source/Core/VideoCommon/VideoConfig.h (hunk 5 above)
Comment (void-ghost) [inline Source/Core/VideoCommon/VideoConfig.h]: Main factor was that I needed three presets myself (normal, low, flat).

File: Source/Core/DolphinWX/Frame.cpp [hunk 8]
```diff
@@ -1416,29 +1416,73 @@ void CFrame::ParseHotkeys()
...
 		if (g_Config.iStereoConvergence > 500)
 			g_Config.iStereoConvergence = 500;
+		g_Config.oStereoPresets[g_Config.iStereoActivePreset].convergence = g_Config.iStereoConvergence;
+		savePreset("Convergence", g_Config.iStereoConvergence);
 	}
 
+	if (IsHotkey(HK_SWITCH_STEREOSCOPY_PRESET))
+	{
+		g_Config.iStereoActivePreset = !g_Config.iStereoActivePreset;
+		g_Config.iStereoConvergence= g_Config.oStereoPresets[g_Config.iStereoActivePreset].convergence;
```
Comment (degasus) [inline Source/Core/DolphinWX/Frame.cpp]: Space before =

File: Source/Core/DolphinWX/Frame.cpp [hunk 9]
```diff
@@ -1416,29 +1416,73 @@ void CFrame::ParseHotkeys()
...
+	if (IsHotkey(HK_INCREASE_CONVERGENCE))
 	{
 		g_Config.iStereoConvergence += 5;
 		if (g_Config.iStereoConvergence > 500)
 			g_Config.iStereoConvergence = 500;
+		g_Config.oStereoPresets[g_Config.iStereoActivePreset].convergence = g_Config.iStereoConvergence;
+		savePreset("Convergence", g_Config.iStereoConvergence);
 	}
 
+	if (IsHotkey(HK_SWITCH_STEREOSCOPY_PRESET))
```
Comment (degasus) [inline Source/Core/DolphinWX/Frame.cpp]: What shall this hotkey do? Switch to 0 if the current preset is != 0? Why shall it switch to exactly 1 if the current present is 0? I think this hotkey is a bit redundant.

File: Source/Core/VideoCommon/VideoConfig.h [hunk 10]
```diff
@@ -84,6 +92,8 @@ struct VideoConfig final
 	int iStereoDepth;
 	int iStereoConvergence;
 	bool bStereoSwapEyes;
+	StereoscopyPreset oStereoPresets[STEREOSCOPY_PRESETS_NUM];
```
Comment (lioncash) [inline Source/Core/VideoCommon/VideoConfig.h]: std::array

**Comment by degasus** [comment]
LGTM

**Comment by phire** [comment]
Sorry, I'm just using this PR to test those random FifoCI failures

@dolphin-emu-bot rebuild

**Comment by eckso** [comment]
How can I turn off this?
For some reason after this merge all my settings behave bad.
As example. Mario Kart before this merge  was perfect with the next settings:

[Video_Enhancements]
StereoConvergence = 442
StereoDepth = 21

**Comment by eckso** [comment]
Why isn't this commit reverted at least while the user works on fixed code?
We shouldn't be green lighting this kind of unasked "features" from fleeting developers, moreover when it's clearly this half-baked. For the time being nobody can use stereoscopic 3D.

**Comment by JMC47** [comment]
I would have made a revert PR myself, but, github says it cannot be automatically reverted or already was.

**Pull request merged**